<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# jsdoc-example-require-spacing

> [ESLint rule][eslint-rules] to enforce empty lines between `require` statements and code in JSDoc examples.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing' );
```

#### rule

[ESLint rule][eslint-rules] to enforce empty lines between `require` statements and code in JSDoc examples.

**Bad**:

<!-- eslint stdlib/jsdoc-example-require-spacing: "off", @cspell/spellchecker: "off" -->

```javascript
/**
* Fréchet distribution constructor.
*
* @module @stdlib/stats/base/dists/frechet/ctor
*
* @example
* var Frechet = require( '@stdlib/stats/base/dists/frechet/ctor' );
* var frechet = new Frechet( 1.0, 1.0, 0.5 );
*
* var y = frechet.cdf( 0.8 );
* // returns ~0.036
*/
```

**Good**:

<!-- eslint @cspell/spellchecker: "off" -->

```javascript
/**
* Fréchet distribution constructor.
*
* @module @stdlib/stats/base/dists/frechet/ctor
*
* @example
* var Frechet = require( '@stdlib/stats/base/dists/frechet/ctor' );
*
* var frechet = new Frechet( 1.0, 1.0, 0.5 );
*
* var y = frechet.cdf( 0.8 );
* // returns ~0.036
*/
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing' );

var linter = new Linter();
var result;
var code;

code = [
    '/**',
    '* Fréchet distribution constructor.',
    '*',
    '* @module @stdlib/stats/base/dists/frechet/ctor',
    '*',
    '* @example',
    '* var Frechet = require( \'@stdlib/stats/base/dists/frechet/ctor\' );',
    '* var frechet = new Frechet( 1.0, 1.0, 0.5 );',
    '*',
    '* var y = frechet.cdf( 0.8 );',
    '* // returns ~0.036',
    '*/',
    'function Frechet() {}'
].join( '\n' );

linter.defineRule( 'jsdoc-example-require-spacing', rule );

result = linter.verify( code, {
    'rules': {
        'jsdoc-example-require-spacing': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'jsdoc-example-require-spacing',
            'severity': 2,
            'message': 'Missing empty line between require statement and code',
            'line': 13,
            'column': 1,
            'nodeType': 'FunctionDeclaration',
            'endLine': 13,
            'endColumn': 21
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
