<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# require-last-path-relative

> [ESLint rule][eslint-rules] enforcing that the last `require` statement is a relative path.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/require-last-path-relative' );
```

#### rule

[ESLint rule][eslint-rules] enforcing that the last `require` statement is a relative path.

**Bad**:

```javascript
var zip = require( '@stdlib/utils/zip' );
var unzip = require( '@stdlib/utils/unzip' );
```

**Good**:

<!-- eslint stdlib/require-file-extensions: "off" -->

```javascript
var zip = require( '@stdlib/utils/zip' );
var unzip = require( './../lib' );
```

The rule is mainly for `examples` files to help make explicit that the API being demonstrated belongs to the respective package.

</section>

<!-- /.usage -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/require-last-path-relative' );

var linter = new Linter();
var result;

var code = [
    'var beep = require( \'@stdlib/boop/beep\' );',
    'var foo = require( \'@stdlib/bar/foo\' );'
].join( '\n' );

linter.defineRule( 'require-last-path-relative', rule );

result = linter.verify( code, {
    'rules': {
        'require-last-path-relative-relative': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'require-last-path-relative',
            'severity': 2,
            'message': 'Last `require` statement must be a relative path',
            'line': 2,
            'column': 11,
            'nodeType': 'CallExpression',
            'endLine': 2,
            'endColumn': 39
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
